src/plugins/ui_actions/public/context_menu/build_eui_context_menu_panels.tsx

```diff
@@ -157,35 +157,51 @@ export async function buildContextMenuForActions({
     const context: ActionExecutionContext<object> = { ...item.context, trigger: item.trigger };
     const isCompatible = await item.action.isCompatible(context);
     if (!isCompatible) return;
-    let parentPanel = '';
-    let currentPanel = '';
+
+    // Reference to the last group...groups are provided in array order of
+    // parent to children (or outer to inner)
+    let parentPanelId = '';
+
     if (action.grouping) {
       for (let i = 0; i < action.grouping.length; i++) {
         const group = action.grouping[i];
-        currentPanel = group.id;
-        if (!panels[currentPanel]) {
+        const groupId = group.id;
+
+        if (!panels[groupId]) {
           const name = group.getDisplayName ? group.getDisplayName(context) : group.id;
-          panels[currentPanel] = {
-            id: currentPanel,
+
+          // Create panel for group
+          panels[groupId] = {
+            id: groupId,
             title: name,
             items: [],
             _level: i,
             _icon: group.getIconType ? group.getIconType(context) : 'empty',
           };
-          if (parentPanel) {
-            panels[parentPanel].items!.push({
+
+          // If there are multiple groups and this is not the first group,
+          // then add an item to the parent panel relating to this group
+          if (parentPanelId) {
+            panels[parentPanelId].items!.push({
               name,
-              panel: currentPanel,
+              panel: groupId,
               icon: group.getIconType ? group.getIconType(context) : 'empty',
               _order: group.order || 0,
               _title: group.getDisplayName ? group.getDisplayName(context) : '',
             });
           }
         }
-        parentPanel = currentPanel;
+
+        // Save the current panel, because this will be used for adding items
+        // to it from later groups in the array
+        parentPanelId = groupId;
       }
     }
-    panels[parentPanel || 'mainMenu'].items!.push({
+
+    // If grouping exists, parentPanelId will be the most-inner group,
+    // otherwise use the mainMenu panel.
+    // Add item for action to this most-inner panel or mainMenu.
+    panels[parentPanelId || 'mainMenu'].items!.push({
```

Comment:
Overall, for this file, yes. We now have borders separating groups, which was a TODO before. There is also some rewrite to make things more readable and with comments to document what is happening, since this functionality is a bit difficult to follow, even with the comments.